src/make_watertight/Arc.cpp

```diff
@@ -0,0 +1,713 @@
+  }
+  std::cout << "error: function not complete" << std::endl;
+  return moab::MB_FAILURE;
+
+  return moab::MB_SUCCESS;
+}
+
+// This function should be rewritten using multimaps or something to avoid
+// upward adjacency searching. Vertices are searched for their adjacent edges.
+moab::ErrorCode Arc::create_loops_from_oriented_edges(moab::Range edges,
+                                                      std::vector< std::vector<moab::EntityHandle> >& loops_of_edges,
+                                                      const bool debug) {
+
+  // conserve edges
+  moab::ErrorCode result;
+  unsigned int n_edges_in  = edges.size();
+  unsigned int n_edges_out = 0;
+  // there could be several arcs for each surface
+  while (0 != edges.size()) {
+    std::vector<moab::EntityHandle> loop_of_edges;
+    // pick initial edge and point
+    moab::EntityHandle edge = edges.front();
+
+    // 20091201 Update: Pinch points may not be important. If not, there is no
+    // purpose detecting them. Instead assume that pinch points coincide with
+    // the endpoints of geometric curves. Also assume that the loop creation at
+    // pinch points does not matter. Pinch points can result in one or more
+    // loops, depending upon the path of traversal through the point.
+
+    // Check to make sure the beginning endpt of the first edge is not a pinch
+    // point. If it is a pinch point the loop is ambiguous. Maybe--see watertightness notes for 20091201
+    {
+      const moab::EntityHandle* end_verts;
+      int n_verts;
+      result = MBI()->get_connectivity(edge, end_verts, n_verts);
+      assert(moab::MB_SUCCESS == result);
+      assert(2 == n_verts);
+      // get the edges adjacent to the back vertex
+      moab::Range adj_edges;
+      result = MBI()->get_adjacencies(&(end_verts[0]), 1, 1, false, adj_edges);
+      assert(moab::MB_SUCCESS == result);
+      // keep the edges that are part of the input range
+      adj_edges = intersect(adj_edges, edges);
+      if (2 != adj_edges.size() && debug) {
+        std::cout << "  create_loops: adj_edges.size()=" << adj_edges.size() << std::endl;
+        std::cout << "  create_loops: pinch point exists" << std::endl;
+        result = MBI()->list_entity(end_verts[0]);
+        assert(moab::MB_SUCCESS == result);
+      }
+    }
+
+    // add it to the loop
+    loop_of_edges.push_back(edge);
+    if (debug)
+      std::cout << "push_back: " << edge << std::endl;
+    n_edges_out++;
+    edges.erase(edge);
+
+    // find connected edges and add to the loop
+    moab::EntityHandle next_edge = 0;
+    while (true) {
+
+      // get the next vertex and next edge
+      result = get_next_oriented_edge(edges, edge, next_edge);
+      if (moab::MB_ENTITY_NOT_FOUND == result) {
+        return result;
+      } else if (moab::MB_SUCCESS != result) {
+        gen->print_arc_of_edges(loop_of_edges);
+        return result;
+      }
+
+      // if the next edge was found
+      if (0 != next_edge) {
```

Comment:
Readability: Rather than doing a `while (true)` loop and a `else break`, might make sense to do a `do` loop with a `while` condition.  E.g.:

```
do {
...
} while (0 != next_edge)
```

Reply:
Personally, I'm ok with the `while(true)`. What I will update here is the if-else to
```
if (next_edge == 0) { break; }
```
which should make it clear where the loop will exit without the additional braces but keep flexibility for other exit points to be added in the future (if needed).